pjrt/convert/trace_container.zig

```diff
@@ -184,12 +224,17 @@ pub const TraceContainer = struct {
         return self;
     }
 
-    pub fn mutableDevice(self: *TraceContainer, allocator: std.mem.Allocator, device_id: u32) !*Device {
-        const entry = try self.metadata.devices.getOrPut(allocator, device_id);
-        if (!entry.found_existing) {
-            entry.value_ptr.* = .{};
+    pub fn getOrPutDevice(self: *TraceContainer, allocator: std.mem.Allocator, key: u32) !GetOrPutResult(u32, ?trace_events_proto.Device) {
```

Comment:
there is a misunderstanding there. This is not what I asked for. 
I want to you to adapt the data structure to the algorithm.

* if you need to insert stuff by key, please use an `AutoArrayHashMap`, and `getOrPutValue`
* if you don't please use an `ArrayList(KV)` and use `.append`. 

The answer to this question probably depend on the thing you are storing. Maybe devices need to use a hash map and arguments an arraylist.

in any case, please explain for each container why you chose this one and not the other.

Reply:
For any protobuf `map<KeyType, ValueType>`, protobuf encodes as a repeated message where the message type is a pair of `KeyType, ValueType`, which is why `zig-protobuf` handles as `ArrayList(KV)`. However, protobuf does not allow duplicate keys, so a better translation would probably be `AutoArrayHashMap`.

Reply:
For `devices`, I've chosen `AutoArrayHashMap` as it enforces deduplication on the key, which is consistent with the protobuf map type/XLA implementation.

For `args`, I've chosen `StringArrayHashMap` as it enforces deduplication on the key, which is consistent with the protobuf map type/XLA implementation. Additionally, the `KV` pairs here are directly outputted as JSON Key/Value pairs; this ensures the resulting output is valid JSON (no duplicate keys).

For `resources`, I've chosen `AutoArrayHashMap` as it enforces deduplication on the key value, which is consistent with the protobuf map type/XLA implementation.